### DISCUSSION TITLE
Data output of grain growth model

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#16190] [Q&A Modules: Phase field] [answered]
Dear MOOSE Experts,

I've started studying the example poly_grain_growth_2D_eldrforce.i.

But I don’t know how to output the evolution of the number of grain boundaries of each grain over time, or the distribution of the number of grain boundaries at each time step.

Any suggestions or recommendations to fix these problems would be greatly appreciated.

Thanks

wei peng

**Comment by aeslaughter** [comment]
@SudiptaBiswas Can you help with this?

**Comment by permcody** [comment, marked as answer]
I'm not sure I understand why you mean by "number of grain boundaries of each grain". Are you wanting the number of grains each grain is adjacent to? If we are on a grain boundary are you interested in the number of grains each boundary is adjacent to? We have a few existing `AuxKernels` that can visualize this information. There is also the [FeataureVolumeVectorPostprocessor](https://mooseframework.inl.gov/moose/source/vectorpostprocessors/FeatureVolumeVectorPostprocessor.html) that can output several "per grain" quantities. If we don't have exactly what you need available now, it shouldn't be difficult to get that information.

**Reply by PengWei97** [reply]
Yes, I want to output the number of neighboring grains for each grain during the time evolution. So what AuxKernel or Postprocessor can achieve this function?

**Reply by permcody** [reply]
Currently this information is not stored. It is saved in the initial condition system (to build the graph coloring matrix), but it's only used in that one instance. However, it is indirectly available during the discovery phase each time the `FeatureFloodCount` object executes, so it could potentially be saved. I don't have time or scope to work on this for you, but I would be willing to advise you on how to modify the source to accomplish this. You could contribute your changes back to MOOSE too. Are you interested in working on this?

**Reply by PengWei97** [reply]
Yeah, I am very honored to contribute to MOOSE. My job is to couple tensor mechanics and phase field to study the process of grain growth, so the grain boundary output is very important to me.

**Reply by permcody** [reply]
Great! So this should be relatively straight forward, but you are going to have to look through the `FeatureFloodCount` object a bit and understand what's there. That object does the bulk of identifying and managing the "feature" or "grain" datastructure. In fact, you'll find the Feature data structure defined in that class.

The code you need for getting a list of all neighbors already exists here: https://github.com/idaholab/moose/blob/b51a7aa8fe7b0665a2ca11b9ec3a313a2c7abaae/modules/phase_field/src/userobjects/PolycrystalUserObjectBase.C#L388-L408
As you may expect, it's a double loop over all the grains to detect "intersecting" grains. In the object above we are doing this so that we can get PETSc to do the "hard" work of generating a coloring map, so I never bothered to store this data in the data structure as this data isn't needed beyond the local scope of where it's used.

So I would proceed like this:
1. Allow the `FeatureFloodCount` object to run it's normal stages like normal (i.e. `initialize`, `execute`, `finalize`). This will populate your data structure with all of the rich information about each feature, it's Halo, and make many of the methods that you'll need to use for your data.
2. In `finalize` I created a new interface that updates data structures as necessary for queries like this one called `updateFieldinfo`. There are already several in there for showing the halos, unique grain maps, etc.  I would recommend that you start by just putting your new logic in that method to perform that double loop and populate a new data structure with a list (or vector) of neighbors for each feature.
3. Copy and paste the code from the PolycrystalUserObject base to populate your new data structure. For debugging, I would recommend that you just print your data right to the terminal. Something like this:
```
grain_id, neighbor_ids
1: 4,5,9
2: 1,3,5
3: 1,2,7,8
...
```
4. Create a new interface in `FeatureVolumeVectorPostprocessor` to get your data back out. I wouldn't recommend that you try to do your own I/O. You are better off using existing MOOSE interfaces. So if you can populate new vectors in this object, you'll get all your data out in the normal MOOSE way: CSV files, Exodus files, etc.

**Reply by permcody** [reply]
BTW - Shameless plug: I have a new publication that should have a preprint available shortly that goes into a lot of the detail of how these algorithms all tie together. Maybe by mid to late December. I'll try to remember to post the link here when I have it available.

**Reply by PengWei97** [reply]
Thank you for your enthusiastic help, and I am very sorry to reply to you now. I will read your suggestions carefully, and hope to achieve what I need. Thank you again for your guidance.

**Reply by PengWei97** [reply]
hi permcody, @permcody 
I want to clarify the information I need to output again in the phase field grain growth simulation. 

[In the simulation of grain growth, the information that needs to be captured is two major aspects. 
](https://www.sciencedirect.com/science/article/pii/S1359645402000848?casa_token=dXzgwCTOyaYAAAAA:oW5pgWUS5-zHc0rGo0rWkRMeDlmJYUK319G6Szif267jiSpPMwhFv_-pAm__cPKjZ7ZP4k4)
- The first is grain-growth kinetics, which is mainly the grain size distribution at each time point. 
- The second is topology of grain-growth-induced microstructures, which needs to be able to output a series of csv files similar to the first one, and each csv file represents the topology of grains at different time nodes.

The content of the topology includes: 

1. the number of faces (sides) per grain, in other words, the number of adjacent grains of a certain grain. I think this is the purpose of writing this object. 
2. the number of edges per grain.
3. the number of vertices (corners) per grain.

According to your suggestion,`you are going to have to look through the FeatureFloodCount object a bit and understand what's there.` I just tried to read the code to understand [FeatureFloodCount](https://github.com/idaholab/moose/blob/next/modules/phase_field/src/postprocessors/FeatureFloodCount.C), but I found this is difficult for beginners like me. So, could you share some basic documents with me, such as the program framework of FeatureFloodCount.C, the entry-level information of this algorithm, etc., so that I can better understand the FeatureFloodCount object.

Finally, I am very sorry to reply you after such a long time. I hope that I could complete the editing of this object with your help and make some contributions to moose, which will also give me a deeper understanding of moose.

Thanks  
wei peng

**Reply by PengWei97** [reply]
Hi @permcody, I think I'm done with outputting the number of adjacent grains by adding some code in FeatureFloodCount and FeatureFloodCount. And I also have made [a pull request](https://github.com/idaholab/moose/pull/20439). Looking forward to your comments.

Wei
